gc/base/standard/Scavenger.cpp

```diff
@@ -1988,12 +1988,65 @@ MM_Scavenger::incrementalScavengeObjectSlots(MM_EnvironmentStandard *env, omrobj
  ****************************************
  */
 
+void
+MM_Scavenger::externalNotifyToYield(MM_EnvironmentBase *env)
+{
+#if defined(OMR_GC_CONCURRENT_SCAVENGER)
+	if (isCurrentPhaseConcurrent()) {
+		omrthread_monitor_enter(_scanCacheMonitor);
+		_shouldYield = true;
+		if (0 != _waitingCount) {
+			omrthread_monitor_notify_all(_scanCacheMonitor);
+		}
+		omrthread_monitor_exit(_scanCacheMonitor);
+	}
+#endif 	/* OMR_GC_CONCURRENT_SCAVENGER */
+}
+
 MMINLINE void
 MM_Scavenger::flushBuffersForGetNextScanCache(MM_EnvironmentStandard *env, bool finalFlush)
 {
 	_delegate.flushReferenceObjects(env);
 	flushRememberedSet(env);
 	flushCopyScanCounts(env, finalFlush);
+
+#if defined(OMR_GC_CONCURRENT_SCAVENGER)
+	if (_extensions->concurrentScavengeExhaustiveTermination && isCurrentPhaseConcurrent()) {
+		/* need to return empty caches to the global pool so they can be accurately counted when evaluating 'all caches returned' termination criteria */
+		returnEmptyCopyCachesToFreeList(env);
+	}
+#endif /* #if defined(OMR_GC_CONCURRENT_SCAVENGER) */
+}
+
+bool
+MM_Scavenger::shouldDoFinalNotify(MM_EnvironmentStandard *env)
+{
+#if defined(OMR_GC_CONCURRENT_SCAVENGER)
+	if (_extensions->concurrentScavengeExhaustiveTermination && isCurrentPhaseConcurrent() && !_scavengeCacheFreeList.areAllCachesReturned()) {
+		
+		/* GC threads ran out of work, but not all copy caches are back to the global free pool. They are kept by mutator threads.
+		 * Activate Async Singnal handler which will force Mutator threads to flush their copy caches for scanning. */
+		_delegate.signalThreadsToFlushCaches(env);
+
+		/* If no work has been created and no one requested to yeild meanwhile, go and wait for new work */
+		if ((0 == _cachedEntryCount) && !checkAndSetShouldYieldFlag(env)) {
+			Assert_MM_true(!_scavengeCacheFreeList.areAllCachesReturned());
```

Comment:
Is it correct that race it is not possible here?

Reply:
This method can be entered by only one thread. That is ensured at the caller site (exisisting functionality) by 1) holding the scan monitor (to prevent othre threads) and 2) threadCount == waiting count  (it's the last one to block).

However there is a nasty race between one GC thread entering this method and another mutator thread trying to acquire Exclusive VM access. Potential risk to create a deadlock since 1) GC (during concurrent phase) holds VM accessses (so mutator thread waits GC to release it), 2) mutator thread has unflushed copy cache (so GC waits for it to flush it). 
To prevent the deadlock: 1) mutator thread will flush caches and notify GC just prior to proceeding to request Exclusive (see changes in EnvBase) and 2) GC thread will do yield check (ie if there is Exclusive request pending) just before starts waiting for mutators to flush (the check just above this comment)